src/patch/apply.js

```diff
@@ -42,7 +44,90 @@ export function applyPatch(source, uniDiff, options = {}) {
             return false;
           }
         }
+        toPos++;
       }
+    }
+
+    return true;
+  }
+
+  function distanceIterator(toPos, minLine, maxLine) {
+    let wantForward = true,
+        backwardExhausted = false,
+        forwardExhausted = false,
+        localOffset = 1;
+
+    return function iterator() {
+      if (wantForward && !forwardExhausted) {
+        if (backwardExhausted) {
+          localOffset++;
+        } else {
+          wantForward = false;
+        }
+
+        // Check if trying to fit beyond text length, and if not, check it fits
+        // after offset location (or desired location on first iteration)
+        if (toPos + localOffset <= maxLine) {
+          return localOffset;
+        }
+
+        forwardExhausted = true;
+      }
+
+      if (!backwardExhausted) {
+        if (!forwardExhausted) {
+          wantForward = true;
+        }
+
+        // Check if trying to fit before text beginning, and if not, check it fits
+        // before offset location
+        if (minLine <= toPos - localOffset) {
+          return -localOffset++;
+        }
+
+        backwardExhausted = true;
+        return iterator();
+      }
+
+      // We tried to fit hunk before text beginning and beyond text lenght, then
+      // hunk can't fit on the text. Return undefined
+    };
+  }
+
+  // Search best fit offsets for each hunk based on the previous ones
+  for (let i = 0; i < hunks.length; i++) {
+    let hunk = hunks[i],
+        maxLine = lines.length - hunk.oldLines,
+        localOffset = 0,
+        toPos = offset + hunk.oldStart - 1;
+
+    let iterator = distanceIterator(toPos, minLine, maxLine);
+
+    for (; localOffset != undefined; localOffset = iterator()) {
```

Comment:
> You need !== here or != null otherwise zero will fail.

```Javascript
[piranna@Mabuk:~/Proyectos/NodeOS/node_modules]
 (vagga) > node
> 0 != undefined
true
> 
```

I think it's not needed... :-) But I could agree on the `!== undefined`, in any case is the only thing it will return... I left it this way to be more generic.


> We should have a test case for that.

A test case for what, exactly?